### ISSUE TITLE
Leverage dictionary-encode when turning a scalar columnar value into an array

### ISSUE THREAD

**Issue by doki23** [apache/datafusion#11503]
### Is your feature request related to a problem or challenge?

We have a `into_array` function of `ColumnarValue` which converts it into an arrow array like this:
```rust
pub fn into_array(self, num_rows: usize) -> Result<ArrayRef> {
      Ok(match self {
          ColumnarValue::Array(array) => array,
          ColumnarValue::Scalar(scalar) => scalar.to_array_of_size(num_rows)?,
      })
  }
```
If this column is of an Int32 type, it returns an Int32Array.
Given that the `ColumnarValue::Scalar`'s cardinality is 1, in some special cases, we can turn it into a dictionary array in order to obtain some performance gains.

### Describe the solution you'd like

For example, if we want to turn a scalar value of f64 type into an array of size 256, we can return a dictionary f64 array of uint8 key type.

**Comment by alamb** [comment]
I agree this is a good idea

It may be easier to take advantage of after @notfilippo  's proposal for logical types https://github.com/apache/datafusion/issues/11513

**Comment by doki23** [comment]
> It may be easier to take advantage of after @notfilippo 's proposal for logical types https://github.com/apache/datafusion/issues/11513

Yes, I also feel a little bit weird that `Dictionary` is a type of arrow. Actually, it's more like an encoding/layout than a data type, right?

**Comment by alamb** [comment]
Update here is there is a more fully fleshed out proposal for logical types: https://github.com/apache/datafusion/issues/11513
